s3-store/src/lib.rs

```diff
@@ -0,0 +1,165 @@
+use std::collections::HashMap;
+use std::io;
+use std::sync::{Arc, Mutex};
+
+use bao_tree::blake3::Hash;
+use bao_tree::io::outboard::PreOrderMemOutboard;
+use bao_tree::{BaoTree, ByteNum, ChunkRanges};
+use bytes::Bytes;
+use futures::future::{self, BoxFuture, LocalBoxFuture};
+use futures::FutureExt;
+use iroh_bytes::store::bao_tree::blake3;
+use iroh_bytes::store::MapEntry;
+use iroh_bytes::IROH_BLOCK_SIZE;
+use iroh_io::{AsyncSliceReader, AsyncSliceReaderExt, HttpAdapter};
+use url::Url;
+
+#[derive(Debug, Clone, Default)]
+pub struct S3Store(Arc<Inner>);
+
+#[derive(Debug, Default)]
+struct Inner {
+    entries: Mutex<HashMap<blake3::Hash, Entry>>,
+}
+
+impl S3Store {
+    pub async fn import_mem(&self, data: Bytes) -> anyhow::Result<Hash> {
+        let size = data.as_ref().len() as u64;
+        let (mut outboard, hash) = bao_tree::io::outboard(&data, IROH_BLOCK_SIZE);
+        outboard.splice(0..8, []);
+        let tree = BaoTree::new(ByteNum(size), IROH_BLOCK_SIZE);
+        let outboard = PreOrderMemOutboard::new(hash, tree, outboard.into())
+            .map_err(|e| anyhow::anyhow!("outboard creation fail {}", e))?;
+        let mut state = self.0.entries.lock().unwrap();
+        state.insert(
+            hash,
+            Entry::new(hash, size, DataDescriptor::Inline(data), outboard),
+        );
+        Ok(hash)
+    }
+
+    pub async fn import_url(&self, url: Url) -> anyhow::Result<Hash> {
+        let mut http_adapter = HttpAdapter::new(url.clone()).await?;
+        let data = http_adapter.read_to_end().await?;
```

Comment:
this reads the full file into memory, right? Can this be streaming instead? To not kill your server when importing a multi-gigabyte file.

Reply:
yeah, bao-tree has helper functions for streaming outboard computation, but only for sync ops.

for async I did not do it yet, so you would have to do this yourself or add the helper fn.